mm/ktsan/shadow.c

```diff
@@ -68,6 +57,18 @@ void ktsan_free_page(struct page *page, unsigned int order)
 	kt_atomic64_fetch_add_no_ktsan(&kt_shadow_pages,
 		-(1 << (order + KT_SHADOW_SLOTS_LOG)));
 
+	page_addr = (unsigned long)page_address(page);
```

Comment:
You are right, it should be fixed in a different way.
Now we always free syncs if memblock exists and always do write to shadow if it is mapped.